### ISSUE TITLE
Add VC metric to track validator statuses

### ISSUE THREAD

**Issue by nflaig** [ChainSafe/lodestar#5158]
**Is your feature request related to a problem? Please describe.**

As a node operator I want to know the status of my validators without relying on 3rd party solutions such as [beaconchai.in](https://beaconcha.in/).

The following questions should be answered

- how many of my validators are active vs just imported without deposit?
- how many validators have a pending deposit?
- how many of my validators exited and was it voluntary or forced?
- what is the state of my withdrawal if exited?

**Describe the solution you'd like**

Add VC metric to track [validator statuses](https://hackmd.io/ofFJ5gOmQpu1jjHilHbdQQ) based on [getValidatorStatus](https://github.com/ChainSafe/lodestar/blob/0607fef8d9b576fd127952122cbf60adc589c629/packages/beacon-node/src/api/impl/beacon/state/utils.ts#L78).

There is already a metric to track total count of validators (`vc_indices_count`), might just need to add `status` label.

https://github.com/ChainSafe/lodestar/blob/0607fef8d9b576fd127952122cbf60adc589c629/packages/validator/src/metrics.ts#L241-L244

This metric should then be visualized on a dashboard, [Lodestar validator client](https://github.com/ChainSafe/lodestar/blob/0607fef8d9b576fd127952122cbf60adc589c629/dashboards/lodestar_validator_client.json) looks like a good candidate as it already displays the VC indices.

Finally, also need to update the metric value used for [`validator_active`](https://github.com/ChainSafe/lodestar/blob/63479b4e9a910c0d0b49fe323d367a12455b7a8a/packages/beacon-node/src/monitoring/clientStats.ts#L185) in the client monitoring to correctly only count the active validators instead of total.

**Comment by philknows** [comment]
This kind of ties into re-thinking a little bit about how the validator UX should look like (#4785) to the end user and how we can improve from our side.

**Comment by maschad** [comment]
Agreed @philknows and I think this can inform how we approach https://github.com/ChainSafe/lodestar/issues/5192

Along with the `status` label we may want to track other metadata like:

- `fee_recipient` set for that validator
- `pubkey`
- `validator index`

**Comment by maschad** [comment]
[Lighthouse's Validator Monitor Metrics](https://github.com/sigp/lighthouse-metrics/blob/master/dashboards/ValidatorMonitor.json#L38) tracks a few things we may be interested in such as:

- [Total Monitored Validators](https://github.com/sigp/lighthouse-metrics/blob/master/dashboards/ValidatorMonitor.json#L111)
- [Active Validators](https://github.com/sigp/lighthouse-metrics/blob/master/dashboards/ValidatorMonitor.json#L207)
- [Slashed Validators](https://github.com/sigp/lighthouse-metrics/blob/master/dashboards/ValidatorMonitor.json#L266)
- [Validator Balance](https://github.com/sigp/lighthouse-metrics/blob/master/dashboards/ValidatorMonitor.json#L332)
- [Exited Validators](https://github.com/sigp/lighthouse-metrics/blob/master/dashboards/ValidatorMonitor.json#L428)
- [Withdrawable Validators](https://github.com/sigp/lighthouse-metrics/blob/master/dashboards/ValidatorMonitor.json#LL487C8-L487C8)

I think this is pretty comprehensive. wdyt @nflaig ?

**Comment by maschad** [comment]
Following up from https://github.com/ChainSafe/lodestar/pull/5161 we would need to do some refactoring in order to get all this data in an accessible place, we want to ensure that we log **every epoch**  i.e. [not conditionally](https://github.com/ChainSafe/lodestar/blob/9b6cab154b81cb573ef2b62df989558346314f1f/packages/validator/src/services/validatorStore.ts#L177) - The main issue here is that given  Validator status  should be logged and tracked **every epoch**, [getStateValidators](https://github.com/ChainSafe/lodestar/blob/9b6cab154b81cb573ef2b62df989558346314f1f/packages/api/src/beacon/routes/beacon/state.ts#L136) , what's the cost of making this api call every epoch? @tuyennhv @dapplion @wemeetagain

**Comment by dapplion** [comment]
Why does it have to be logged every epoch? Where does this requirement come from? getStateValidators is somewhat expensive to keep polling regularly

**Comment by maschad** [comment]
> Why does it have to be logged every epoch? Where does this requirement come from? 

https://github.com/ChainSafe/lodestar/issues/4785#issuecomment-1348892078

That was the impetus behind #5161.

Perhaps we should cache this response then.

**Comment by nflaig** [comment]
The Lighthouse validator monitor dashboard looks pretty complete to me and should answer most of the questions mentioned in the issue.

The question is where should those metrics be gathered from and displayed.  If we use the `validator_monitor` metrics we need to get them from the BN and display them on the [validator monitor dashboard](https://github.com/ChainSafe/lodestar/blob/unstable/dashboards/lodestar_validator_monitor.json).

The initial idea of the issue was to get the metrics from the VC and display them on the [validator client dashboard](https://github.com/ChainSafe/lodestar/blob/unstable/dashboards/lodestar_validator_client.json). In the end, I think we want both as we should see the VC and BN as independent services.

**Comment by maschad** [comment]
I would say if we are using Lighthouse as an example it would be more appropriate in the [validator **monitor** dashboard](https://github.com/ChainSafe/lodestar/blob/unstable/dashboards/lodestar_validator_monitor.json?rgh-link-date=2023-03-03T07%3A49%3A51Z) , I think we can delegate metrics related to attestations, publishing, block proposals, proofs and overall consensus between nodes to the [validator **client** dashboard](https://github.com/ChainSafe/lodestar/blob/unstable/dashboards/lodestar_validator_client.json?rgh-link-date=2023-03-03T07%3A49%3A51Z).

**Comment by nflaig** [comment]
They still have few metrics in their [validator client dashboard](https://github.com/sigp/lighthouse-metrics/blob/master/dashboards/ValidatorClient.json), such as [total validators](https://github.com/sigp/lighthouse-metrics/blob/71054d58b340a5f01f0da0cc24f900035247388f/dashboards/ValidatorClient.json#L908) and [enabled validators](https://github.com/sigp/lighthouse-metrics/blob/71054d58b340a5f01f0da0cc24f900035247388f/dashboards/ValidatorClient.json#L807).

**Comment by wemeetagain** [comment]
> Perhaps we should cache this response then.

This won't do if we're trying to track statuses. These statuses are things that can change every epoch, so using cached statuses will just return the old statuses and not the (possibly) changed statuses.

>  getStateValidators is somewhat expensive to keep polling regularly

@dapplion what do you mean? we poll it regularly right now (as long as we haven't discovered all attached validators)

RE validator monitor vs validator client metrics, my opinion is that additional logs / metrics is a nice-to-have that should be weighed against additional validator / beacon node load. We also need to be very thoughtful and explicit about what additional logs and metrics are actually helpful for users, and what things are just us doing make-work.

**Comment by dapplion** [comment]
> @dapplion what do you mean? we poll it regularly right now (as long as we haven't discovered all attached validators)

Right but once everyone is discovered we stop. If I understand the plan is to check the status of each connected validator every epoch

**Comment by nflaig** [comment]
> Why does it have to be logged every epoch? Where does this requirement come from? getStateValidators is somewhat expensive to keep polling regularly

Looks like all other VCs poll `getStateValidators` every epoch, I agree that should be conservative where possible but if another VC is used with Lodestar BN this will happen either way

**Comment by philknows** [comment]
Benefits: While VC's running, if VCs get exited, you can remove the keys. It will prevent inaccuracies in `active` validator keys.

Edge case: If a validator is in a sync committee as it is exiting.

**Comment by philknows** [comment]
Can we measure the performance impact of the UX improvement to justify polling every epoch? There have been improvements to the interaction since this issue was created, so we can make a case to test this with Vero since they poll every epoch already.

**Comment by nflaig** [comment]
> > Why does it have to be logged every epoch? Where does this requirement come from? getStateValidators is somewhat expensive to keep polling regularly
> 
> Looks like all other VCs poll `getStateValidators` every epoch, I agree that should be conservative where possible but if another VC is used with Lodestar BN this will happen either way

we shouldn't do this unless there is a ssz variant of the api
